### ISSUE TITLE
Build failure: phpPackages.psysh

### ISSUE THREAD

**Issue by lucc** [NixOS/nixpkgs#337952]
### Steps To Reproduce

1. `nix build nixpkgs#phpPackages.psysh`

### Build log

```
error: hash mismatch in fixed-output derivation '/nix/store/qbcxs22v1zhdz2j9bla4yac0qmg92anp-composer.lock.drv':
         specified: sha256-ur6mzla3uXeFL6aEHAPdpxGdvcgzOgTLW/CKPbNqeCg=
            got:    sha256-PQDWShzvTY8yF+OUPVJAV0HMx0/KnA03TDhZUM7ppXw=
error: 1 dependencies of derivation '/nix/store/gdhc1ya9ijnyxbc062mn8h5azdixs2ch-psysh-0.12.4.drv' failed to build
```

### Notify maintainers

@aanderse @drupol @Ma27 @piotrkwiecinski @talyz

(these are the explicit maintainers from `nix eval nixpkgs/master\#phpPackages.psysh.meta.maintainers --apply 'builtins.map ({github, ... }: github)'` because I do not know how to ping the php team. Sorry.)

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.45, NixOS, 24.11 (Vicuna), 24.11.20240814.c3aa7b8`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.5`
 - nixpkgs: `/nix/store/sfycwi72zfjsspidinx56ajaiffpyh17-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by lucc** [comment]
I noticed that psysh is also out of date. There are some new versions available so maybe we can just update the derivation to  a new version.

https://github.com/bobthecow/psysh/releases

**Comment by drupol** [comment]
Not until we have fixed the instability with `buildComposerProject2`.

**Comment by piotrkwiecinski** [comment]
It's related to use of a new builder introduced in https://github.com/NixOS/nixpkgs/pull/308059. Similar to https://github.com/NixOS/nixpkgs/pull/336145. There is an edge case in composer autoloader logic leading to slightly different file order. Here is some discussion https://github.com/composer/composer/pull/12090 but it's not a proper solution yet.

I hope to spend some time later this week to dig deeper into composer internals.

**Comment by lucc** [comment]
This seems to be fixed. I can `nix build` and `nix run` any of these:
- nixpkgs/nixos-24.11#phpPackages.psysh
- nixpkgs/nixos-unstable#phpPackages.psysh
- nixpkgs/master#phpPackages.psysh

**Comment by piotrkwiecinski** [comment]
Yes it is. It was a composer bug https://github.com/composer/composer/pull/12263.
